### testing the XGBoost Diabetes example to transpile
##### used conda env giza from ll laptop

## Create and Train an XGBoost Model
### We'll start by creating a simple XGBoost model using Scikit-Learn and train it on diabetes dataset

In [9]:
import numpy as np
import pandas as pd
import xgboost as xgb
# import matplotlib.pyplot as plt
# import seaborn as sns
# import xgboost as xgb

In [3]:
df = pd.read_csv('BETF Final.csv')
df.head(10)

Date  Closing Price   IBIT   FBTC   BITB   ARKB  BTCO  EZBC  BRRR  \
0  11-01-2024    46368.58594  111.7  227.0  237.9   65.3  17.4  50.1  29.4   
1  12-01-2024    42853.16797  386.0  195.3   17.4   39.8  28.4   0.0  20.2   
2  16-01-2024    43154.94531  212.7  102.0   50.2  122.3  31.9   0.0  15.3   
3  17-01-2024    42742.65234  371.4  358.1   68.2   50.3  57.6   1.2   1.2   
4  18-01-2024    41262.05859  145.5  177.9   20.1   41.8  58.8   0.0   9.3   
5  19-01-2024    41618.40625  201.5  222.3   56.7   62.6  63.4   0.0  10.4   
6  22-01-2024    39507.36719  260.6  158.7   41.6   65.0   5.6   4.7   9.7   
7  23-01-2024    39845.55078  160.1  157.7   26.3   61.8   0.0   1.1   0.0   
8  24-01-2024    40077.07422   66.2  125.7   19.1   24.9  19.9   1.2   9.1   
9  25-01-2024    39933.80859  170.7  101.0   20.0   16.1   0.0   0.0   6.5   

   HODL  BTCW   GBTC  Total  
0  10.6   1.0  -95.1  655.3  
1   0.0   0.0 -484.1  203.0  
2   7.3   0.0 -594.4  -52.7  
3   4.8   1.6 -460.6  453.8  
4   2.3   0.0 -582.3 -126.6  
5  14.2   2.9 -590.4   43.6  
6   6.8   0.4 -640.5  -87.4  
7   2.2   0.0 -515.3 -106.1  
8   4.5   0.4 -429.3 -158.3  
9   0.0   0.0 -394.1  -79.8

In [4]:
y = df['Closing Price']

In [5]:
X = df[['IBIT', 'FBTC', 'BITB', 'ARKB', 'BTCO',
       'EZBC', 'BRRR', 'HODL', 'BTCW',
       'GBTC', ]]

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=410)

In [7]:
from sklearn.linear_model import LinearRegression

In [10]:
xgb_model = xgb.XGBRegressor(n_estimators=100, random_state=410)
xgb_model.fit(X_train, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=100, n_jobs=None,
             num_parallel_tree=None, random_state=410, ...)

In [11]:
predictions = xgb_model.predict(X_test)
predictions

array([57026.3  , 57109.89 , 68121.38 , 46540.348, 61123.418, 61416.383,
       52103.418, 62789.984, 61848.367, 62689.957, 61015.64 , 62924.355,
       56717.832, 61962.96 , 59422.047, 51762.324, 62287.754, 66328.1  ,
       45815.62 , 68623.61 , 65385.88 , 65655.62 , 44231.605, 60906.49 ,
       62002.926, 40723.25 , 59347.02 , 68672.375, 62167.168, 68079.04 ],
      dtype=float32)

In [12]:
# import xgboost as xgb
# from sklearn.datasets import load_diabetes
# from sklearn.model_selection import train_test_split

# data = load_diabetes()
# X, y = data.data, data.target

# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# # Increase the number of trees and maximum depth
# n_estimators = 2  # Increase the number of trees
# max_depth = 6  # Increase the maximum depth of each tree

# xgb_reg = xgb.XGBRegressor(n_estimators=n_estimators, max_depth=max_depth)
# xgb_reg.fit(X_train, y_train)

## Save the model
### Save the model in Json format

In [13]:
from giza.zkcook import serialize_model
serialize_model(xgb_model, "etf_xgb1.json")

## Transpile your model to Orion Cairo
### We will use Giza-CLI to transpile our saved model to Orion Cairo.


$ giza transpile etf_xgb1.json --output-path etf_xgb1
[giza][2024-06-03 19:00:12.216] No model id provided, checking if model exists ✅
[giza][2024-06-03 19:00:12.229] Model name is: etf_xgb1
[giza][2024-06-03 19:00:12.813] Model Created with id -> 690! ✅
[giza][2024-06-03 19:00:13.400] Version Created with id -> 1! ✅
[giza][2024-06-03 19:00:13.413] Sending model for transpilation ✅
[giza][2024-06-03 19:01:06.244] Transpilation is fully compatible. Version compiled and Sierra is saved at Giza ✅
[giza][2024-06-03 19:01:07.583] Downloading model ✅
[giza][2024-06-03 19:01:07.594] model saved at: etf_xgb1



## Deploy an inference endpoint
### Now that our model is transpiled to Cairo we can deploy an endpoint to run verifiable inferences. We will use Giza CLI again to run and deploy an endpoint. Ensure to replace model-id and version-id with your ids provided during transpilation.


$ giza endpoints deploy --model-id 690 --version-id 1
▰▰▰▰▱▱▱ Creating endpoint!
[giza][2024-06-03 19:03:08.874] Endpoint is successful ✅
[giza][2024-06-03 19:03:08.880] Endpoint created with id -> 261 ✅
[giza][2024-06-03 19:03:08.882] Endpoint created with endpoint URL: https://endpoint-giza1-690-1-62d762c5-7i3yxzspbq-ew.a.run.app 🎉

### Create and Agent using the CLI

$ giza agents create --model-id 690 --version-id 1 --name etfXGB --description etfXGB
[giza][2024-06-03 19:11:50.498] Creating agent ✅
[giza][2024-06-03 19:11:50.506] Using model id and version id to create agent
[giza][2024-06-03 19:11:50.780] Select an existing account to create the agent.
[giza][2024-06-03 19:11:50.787] Available accounts are:
┌──────────┐
│ Accounts │
├──────────┤
│  giza1   │
└──────────┘
Enter the account name: giza1
{
  "id": 34,
  "name": "etfXGB",
  "description": "etfXGB",
  "parameters": {
    "model_id": 690,
    "version_id": 1,
    "endpoint_id": 261,
    "account": "giza1"
  },
  "created_date": "2024-06-04T02:11:57.598948",
  "last_update": "2024-06-04T02:11:57.598948"
}

In [30]:
X_test

array([[ 0.04534098, -0.04464164, -0.00620595, -0.01599898,  0.1250187 ,
         0.1251981 ,  0.019187  ,  0.03430886,  0.03243232, -0.0052198 ],
       [ 0.09256398, -0.04464164,  0.03690653,  0.02187239, -0.02496016,
        -0.01665815,  0.00077881, -0.03949338, -0.02251653, -0.02178823],
       [ 0.06350368,  0.05068012, -0.00405033, -0.01255612,  0.10300346,
         0.04878988,  0.05600338, -0.00259226,  0.08449153, -0.01764613],
       [ 0.09619652, -0.04464164,  0.0519959 ,  0.07926471,  0.05484511,
         0.03657709, -0.07653559,  0.14132211,  0.09864806,  0.06105391],
       [ 0.01264814,  0.05068012, -0.02021751, -0.00222757,  0.03833367,
         0.05317395, -0.00658447,  0.03430886, -0.00514219, -0.00936191],
       [ 0.0090156 , -0.04464164, -0.02452876, -0.02632753,  0.0988756 ,
         0.0941964 ,  0.07072993, -0.00259226, -0.02139531,  0.00720652],
       [-0.00914709,  0.05068012,  0.17055523,  0.01498668,  0.03007796,
         0.03375875, -0.02131102,  0.03430886

In [27]:
new_data = pd.DataFrame({
    'IBIT': [120],
    'FBTC': [175],
    'BITB': [5],
    'ARKB': [50],
    'BTCO': [1],
    'EZBC': [2],
    'BRRR': [3],
    'HODL': [4],
    'BTCW': [5],
    'GBTC': [-20]
})

In [28]:
predictions = xgb_model.predict(X_test)
predictions

array([62310.82 , 62310.82 , 62310.78 , 62310.82 , 62310.78 , 62310.82 ,
       60597.242, 60597.242, 60597.242, 60597.242, 60597.28 , 62310.82 ,
       60597.28 , 62310.78 , 60597.242, 60597.242, 62310.78 , 62310.78 ,
       60597.242, 62310.78 , 60597.28 , 62310.82 , 60597.28 , 62310.78 ,
       60597.28 , 62310.78 , 62310.78 , 62310.78 , 60597.28 , 62310.78 ,
       62310.82 , 60597.28 , 60597.242, 62310.78 , 62310.82 , 60597.242,
       60597.242, 60597.242, 62310.78 , 62310.82 , 60597.28 , 62310.78 ,
       62310.78 , 62310.78 , 62310.82 , 60597.28 , 60597.28 , 62310.82 ,
       60597.28 , 60597.28 , 62310.82 , 60597.28 , 62310.78 , 62310.82 ,
       60597.242, 60597.28 , 60597.28 , 62310.78 , 60597.28 , 60597.28 ,
       60597.28 , 60597.28 , 60597.242, 62310.78 , 60597.28 , 62310.82 ,
       60597.242, 62310.78 , 60597.28 , 60597.242, 60597.242, 60597.242,
       62310.82 , 60597.242, 62310.82 , 62310.78 , 60597.242, 60597.28 ,
       62310.78 , 62310.82 , 60597.28 , 60597.28 , 

In [29]:
predictions1 = xgb_model.predict(new_data)
predictions1

array([56899.145], dtype=float32)

array([62310.82 , 62310.82 , 62310.78 , 62310.82 , 62310.78 , 62310.82 ,
       60597.242, 60597.242, 60597.242, 60597.242, 60597.28 , 62310.82 ,
       60597.28 , 62310.78 , 60597.242, 60597.242, 62310.78 , 62310.78 ,
       60597.242, 62310.78 , 60597.28 , 62310.82 , 60597.28 , 62310.78 ,
       60597.28 , 62310.78 , 62310.78 , 62310.78 , 60597.28 , 62310.78 ,
       62310.82 , 60597.28 , 60597.242, 62310.78 , 62310.82 , 60597.242,
       60597.242, 60597.242, 62310.78 , 62310.82 , 60597.28 , 62310.78 ,
       62310.78 , 62310.78 , 62310.82 , 60597.28 , 60597.28 , 62310.82 ,
       60597.28 , 60597.28 , 62310.82 , 60597.28 , 62310.78 , 62310.82 ,
       60597.242, 60597.28 , 60597.28 , 62310.78 , 60597.28 , 60597.28 ,
       60597.28 , 60597.28 , 60597.242, 62310.78 , 60597.28 , 62310.82 ,
       60597.242, 62310.78 , 60597.28 , 60597.242, 60597.242, 60597.242,
       62310.82 , 60597.242, 62310.82 , 62310.78 , 60597.242, 60597.28 ,
       62310.78 , 62310.82 , 60597.28 , 60597.28 , 

## Run a verifiable inference
##### To streamline verifiable inference, you might consider using the endpoint URL obtained after transpilation. However, this approach requires manual serialization of the input for the Cairo program and handling the deserialization process. To make this process more user-friendly and keep you within a Python environment, we've introduced a Python SDK designed to facilitate the creation of ML workflows and execution of verifiable predictions. When you initiate a prediction, our system automatically retrieves the endpoint URL you deployed earlier, converts your input into Cairo-compatible format, executes the prediction, and then converts the output back into a numpy object. 

In [17]:
X_test[1, :]

array([ 0.09256398, -0.04464164,  0.03690653,  0.02187239, -0.02496016,
       -0.01665815,  0.00077881, -0.03949338, -0.02251653, -0.02178823])

In [32]:
import xgboost as xgb
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

from giza.agents.model import GizaModel


MODEL_ID = 690  # Update with your model ID
VERSION_ID = 1  # Update with your version ID

def prediction(input, model_id, version_id):
    model = GizaModel(id=model_id, version=version_id)

    (result, proof_id) = model.predict(
        input_feed={"input": input}, verifiable=True, model_category="XGB"
    )

    return result, proof_id


def execution():
    # The input data type should match the model's expected input
    # input = X_test[1, :]
    input = X_test

    new_data = pd.DataFrame({
    'IBIT': [120],
    'FBTC': [175],
    'BITB': [5],
    'ARKB': [50],
    'BTCO': [1],
    'EZBC': [2],
    'BRRR': [3],
    'HODL': [4],
    'BTCW': [5],
    'GBTC': [-20]
})

    (result, proof_id) = prediction(input, MODEL_ID, VERSION_ID)
    # (result, proof_id) = prediction(new_data, MODEL_ID, VERSION_ID)

    
    # print(f"Predicted value for input {new_data.flatten()[0]} is {result}")
    print(f"Predicted value for input {input.flatten()[0]} is {result}")

    return result, proof_id


if __name__ == "__main__":
    # data = load_diabetes()
    # X, y = data.data, data.target

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )
    _, proof_id = execution()
    print(f"Proof ID: {proof_id}")

An error occurred in predict: 503 Server Error: Service Unavailable for url: https://endpoint-giza1-690-1-62d762c5-7i3yxzspbq-ew.a.run.app/cairo_run
Deployment predict error: Service Unavailable
An error occurred in predict: 503 Server Error: Service Unavailable for url: https://endpoint-giza1-690-1-62d762c5-7i3yxzspbq-ew.a.run.app/cairo_run


HTTPError: 503 Server Error: Service Unavailable for url: https://endpoint-giza1-690-1-62d762c5-7i3yxzspbq-ew.a.run.app/cairo_run

## Download the proof
#### Initiating a verifiable inference sets off a proving job on our server, sparing you the complexities of installing and configuring the prover yourself. Upon completion, you can download your proof.

First, let's check the status of the proving job to ensure that it has been completed.

$ giza endpoints get-proof --endpoint-id 234 --proof-id "10c164e6c2364ab6b5491702127979a6"
[giza][2024-05-30 00:40:39.691] Getting proof from endpoint 234 ✅
{
  "id": 967,
  "job_id": 1121,
  "metrics": {
    "proving_time": 17.249508
  },
  "created_date": "2024-05-30T07:33:12.532659"
}

Once the proof is ready, you can download it.

$ giza endpoints download-proof --endpoint-id 234 --proof-id "10c164e6c2364ab6b5491702127979a6" --output-path zk_xgboost.proof
[giza][2024-05-30 00:51:52.048] Getting proof from endpoint 234 ✅
[giza][2024-05-30 00:51:53.800] Proof downloaded to zk_xgboost.proof ✅
(giza3)

## Verify the proof
#### Finally, you can verify the proof.

$ giza verify --proof-id 967
[giza][2024-05-30 00:56:05.847] Verifying proof...
[giza][2024-05-30 00:56:07.140] Verification result: True
[giza][2024-05-30 00:56:07.145] Verification time: 0.454667226
(giza3)